frontend/components/ModelForm.tsx

```diff
@@ -0,0 +1,515 @@
+import { Form, Formik } from 'formik'
+import { Component, ReactElement } from 'react'
+import s from 'styled-components'
+
+import {
+  doApiRequest,
+  formatResponse,
+  getApiUrl,
+  getRoleDisplay,
+  titleize,
+} from '../utils'
+import { Icon, Loading } from './common'
+import { FormFieldClassContext } from './FormComponents'
+
+const ModelItem = s.div`
+  padding: 15px;
+  border: 1px solid #dbdbdb;
+  border-radius: 3px;
+  margin-bottom: 1em;
+`
+
+const ModelStatusWrapper = s.span`
+  display: inline-block;
+  margin: 0.375em 0.75em;
+`
+
+const ModelStatus = ({ status }) => (
+  <ModelStatusWrapper>
+    {typeof status !== 'undefined' &&
+      (status === true ? (
+        <span style={{ color: 'green' }}>
+          <Icon name="check-circle" alt="success" /> Saved!
+        </span>
+      ) : (
+        <span style={{ color: 'red' }}>
+          <Icon name="x-circle" alt="failure" /> Failed to save!
+        </span>
+      ))}
+  </ModelStatusWrapper>
+)
+
+const Subtitle = s.div`
+  font-weight: bold;
+  font-size: 1.2em;
+  margin-bottom: 0.75em;
+`
+
+type ModelFormProps = any
+
+type ModelFormState = any
+
+export const doFormikInitialValueFixes = (currentObject: {
+  [key: string]: any
+}): { [key: string]: any } => {
+  return Object.entries(currentObject).reduce((prev, [key, val]) => {
+    if (key.endsWith('_url')) {
+      const otherKey = key.substr(0, key.length - 4)
+      if (!(otherKey in prev)) {
+        prev[otherKey] = val
+      }
+    }
+    prev[key] = val
+    return prev
+  }, {})
+}
+
+/*
+ * Creates a form with CRUD (create, read, update, delete)
+ * capabilities for a Django model using a provided endpoint.
+ */
+export class ModelForm extends Component<ModelFormProps, ModelFormState> {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      objects: null,
+      currentlyEditing: null,
+      createObject:
+        props.defaultObject != null ? { ...props.defaultObject } : {},
+      newCount: 0,
+    }
+
+    this.onSubmit = this.onSubmit.bind(this)
```

Comment:
if you're using class components, can get around the bind issue by using lambda syntax! They take on the `this` from the parent automatically.

Reply:
This is a neat trick! Thanks for letting me know about this.